src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -1334,6 +1496,21 @@ otError RadioSpinel<InterfaceType, ProcessContextType>::Set(spinel_prop_key_t aK
     va_end(mPropertyArgs);
     mExpectedCommand = SPINEL_CMD_NOOP;
 
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_ENABLE
+    if (mRcpFailed)
+    {
+        RecoverFromRcpFailure();
+
+        mExpectedCommand = SPINEL_CMD_PROP_VALUE_IS;
+        va_start(mPropertyArgs, aFormat);
+        error = RequestV(true, SPINEL_CMD_PROP_VALUE_SET, aKey, aFormat, mPropertyArgs);
+        va_end(mPropertyArgs);
+        mExpectedCommand = SPINEL_CMD_NOOP;
+
+        VerifyOrDie(!mRcpFailed, OT_EXIT_FAILURE);
```

Comment:
```suggestion
```

If we set properties in Spinel properly before calling `Set`, we actually don't need to retry the `Set` operation?
The device properties should be properly restored in `RecoverFromRcpFailure`.

Reply:
The stored properties are updated only after `Set` returns, so the values restored are old ones.

Reply:
> The stored properties are updated only after `Set` returns, so the values restored are old ones.

Can we update stored properties before `Set` ? This should produce simpler code.

Reply:
Then we have to revert the value back if the actual operation fails. Seems more complex.

Reply:
Why do we need to revert? If the operation failed, shouldn't we restore RCP failure then?
If the restoration is successful, we are good. If not, we die. Setting value before actual `Set` removes the necessity to retry `Set`. Thought?

Reply:
The operation may fail because of other reasons, for example, the RCP reports an error.

Reply:
I can see your point. SGTM. 

But another thought is that since the RCP device is under total control of Host OpenThread, we should expect no failure if RCP is working well. Any error from RCP should indicate a device failure and thus should be fixed by RCP Restoration.